nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+    }
+    List<Type> lhsTypeArguments = lhsType.getTypeArguments();
+    List<Type> rhsTypeArguments = rhsType.getTypeArguments();
+    if (lhsTypeArguments.size() != rhsTypeArguments.size()) {
+      throw new RuntimeException(rhsType + " does not match " + lhsType);
+    }
+    for (int i = 0; i < lhsTypeArguments.size(); i++) {
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsLHS =
+          lhsTypeArguments.get(i).getAnnotationMirrors();
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsRHS =
+          rhsTypeArguments.get(i).getAnnotationMirrors();
+      boolean isLHSNullableAnnotated =
+          Nullness.hasNullableAnnotation(annotationMirrorsLHS.stream(), config);
+      boolean isRHSNullableAnnotated =
+          Nullness.hasNullableAnnotation(annotationMirrorsRHS.stream(), config);
+      if (isLHSNullableAnnotated != isRHSNullableAnnotated) {
+        invalidAssignmentInstantiationError(tree, lhsType, rhsType, state, analysis);
+        return;
+      }
+      // nested generics
+      if (lhsTypeArguments.get(i).getTypeArguments().length() > 0) {
+        compareAnnotations(
+            (Type.ClassType) lhsTypeArguments.get(i),
+            (Type.ClassType) rhsTypeArguments.get(i),
+            tree);
+      }
+    }
+  }
+
+  /**
+   * @param tree A parameterized typed tree for which we need class type with preserved annotations.
+   * @return A Type with preserved annotations.
+   */
+  private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree) {
+    Type.ClassType type = (Type.ClassType) ASTHelpers.getType(tree);
+    Preconditions.checkNotNull(type);
+    Type nullableType = NULLABLE_TYPE_SUPPLIER.get(state);
+    List<? extends Tree> typeArguments = tree.getTypeArguments();
+    List<Type> newTypeArgs = new ArrayList<>();
+    for (int i = 0; i < typeArguments.size(); i++) {
+      List<Attribute.TypeCompound> allAnnotations = new ArrayList<>();
+      List<JCTree.JCAnnotation> annotations = new ArrayList<>();
+      if (typeArguments.get(i).getClass().equals(JCTree.JCAnnotatedType.class)) {
+        JCTree.JCAnnotatedType annotatedType = (JCTree.JCAnnotatedType) typeArguments.get(i);
+        annotations = annotatedType.getAnnotations();
+      } else if (typeArguments.get(i) instanceof JCTree.JCTypeApply
+          && ((JCTree.JCTypeApply) typeArguments.get(i)).clazz instanceof AnnotatedTypeTree) {
+        JCTree.JCAnnotatedType annotatedType =
+            (JCTree.JCAnnotatedType) ((JCTree.JCTypeApply) typeArguments.get(i)).clazz;
+        annotations = annotatedType.getAnnotations();
+      }
+      for (JCTree.JCAnnotation annotation : annotations) {
+        Attribute.Compound attribute = annotation.attribute;
+        if (attribute.toString().equals(NULLABLE_TYPE)) {
+          allAnnotations.add(
+              new Attribute.TypeCompound(nullableType, com.sun.tools.javac.util.List.nil(), null));
```

Comment:
`allAnnotations` should only ever have 0 or 1 elements, right?  If I understand correctly, then get rid of `allAnnotations`.  Instead, just track a boolean `hasNullableAnnotation`.  Then below, conditionally assign `annos`: if `hasNullableAnnotation` is true, assign `com.sun...List.from(Collections.singletonList(new Attribute.TypeCompound(...)))`, otherwise assign `com.sun...List.nil()`